test/conformance/helpers/broker_tracing_test_helper.go

```diff
@@ -0,0 +1,489 @@
+	"knative.dev/eventing/pkg/apis/eventing/v1alpha1"
+	"knative.dev/eventing/test/base/resources"
+	"knative.dev/eventing/test/common"
+	tracinghelper "knative.dev/eventing/test/conformance/helpers/tracing"
+	"knative.dev/pkg/test/zipkin"
+)
+
+func BrokerTracingTestHelper(t *testing.T, channelTestRunner common.ChannelTestRunner) {
+	testCases := map[string]struct {
+		incomingTraceId bool
+		istio           bool
+	}{
+		"includes incoming trace id": {
+			incomingTraceId: true,
+		},
+	}
+
+	for n, tc := range testCases {
+		loggerPodName := "logger"
+		t.Run(n, func(t *testing.T) {
+			channelTestRunner.RunTests(t, common.FeatureBasic, func(st *testing.T, channel string) {
+				// Don't accidentally use t, use st instead. To ensure this, shadow 't' to a useless
+				// type.
+				t := struct{}{}
+				_ = fmt.Sprintf("%s", t)
+
+				client := common.Setup(st, true)
+				defer common.TearDown(client)
+
+				// Label namespace so that it creates the default broker.
+				if err := client.LabelNamespace(map[string]string{"knative-eventing-injection": "enabled"}); err != nil {
+					st.Fatalf("Error annotating namespace: %v", err)
+				}
+
+				// Do NOT call zipkin.CleanupZipkinTracingSetup. That will be called exactly once in
+				// TestMain.
+				tracinghelper.Setup(st, client)
+
+				expected, mustContain := setupBrokerTracing(st, channel, client, loggerPodName, tc.incomingTraceId)
+				assertLogContents(st, client, loggerPodName, mustContain)
+				traceID := getTraceID(st, client, loggerPodName)
+				trace, err := zipkin.JSONTrace(traceID, expected.SpanCount(), 2*time.Minute)
+				if err != nil {
+					st.Fatalf("Unable to get trace %q: %v. Trace so far %+v", traceID, err, tracinghelper.PrettyPrintTrace(trace))
+				}
+
+				tree := tracinghelper.GetTraceTree(st, trace)
+				if err := expected.Matches(tree); err != nil {
+					st.Fatalf("Trace Tree did not match expected: %v", err)
+				}
+			})
+		})
+	}
+}
+
+// setupBrokerTracing is the general setup for TestBrokerTracing. It creates the following:
+// 1. Broker.
+// 2. Trigger on 'foo' events -> K8s Service -> transformer Pod (which replies with a 'bar' event).
+// 3. Trigger on 'bar' events -> K8s Service -> eventdetails Pod.
+// 4. Sender Pod which sends a 'foo' event.
+// It returns a string that is expected to be sent by the SendEvents Pod and should be present in
+// the LogEvents Pod logs.
+func setupBrokerTracing(t *testing.T, channel string, client *common.Client, loggerPodName string, incomingTraceId bool) (tracinghelper.TestSpanTree, string) {
+	// Create the Broker.
+	const (
+		brokerName    = "br"
+		etTransformer = "transformer"
+		etLogger      = "logger"
+	)
+	channelTypeMeta := common.GetChannelTypeMeta(channel)
+	client.CreateBrokerOrFail(brokerName, channelTypeMeta)
+
+	// Due to https://github.com/knative/eventing/issues/1998, wait for the Broker to become ready
+	// before creating the Triggers.
+	err := client.WaitForResourceReady(brokerName, &metav1.TypeMeta{
```

Comment:
We can use `common.BrokerTypeMeta` here. 
Also sounds like this is a bug and we'll fix it anyway, can we change the comment to a `TODO`? Since we also have the similar logic in e2e tests.

Reply:
Done.

I think the other e2e tests use the default Broker, which is likely created via the injection annotation and happens to be ready in time.

Reply:
There are also two test cases that create brokers:
https://github.com/knative/eventing/blob/b2e44b5cb15e6ac101ccbea2558e08acca2ec308/test/e2e/helpers/broker_channel_flow_test_helper.go#L62
https://github.com/knative/eventing/blob/b2e44b5cb15e6ac101ccbea2558e08acca2ec308/test/e2e/helpers/broker_event_transformation_test_helper.go#L57